### PR TITLE
【Hackathon 5 No.2】Add index_fill / index_fill_ API to Paddle -part

### PR DESCRIPTION
**Pull request by Patrick-Star125**
### PR types

New features

### PR changes

APIs

### Description

Add index_fill/index_fill_ API to Paddle

### Link

Rfc PR: https://github.com/PaddlePaddle/community/pull/621
docs PR: https://github.com/PaddlePaddle/docs/pull/6284

待完成

- [x] RFC文档修改
- [x] 中文文档

### CODE REVIEW

File: python/paddle/tensor/manipulation.py [hunk 1]
```diff
@@ -5166,3 +5166,107 @@ def unfold(x, axis, size, step, name=None):
...
+    perm[axis] = 0
+
+    out = paddle.clone(x)
+    out = paddle.transpose(out, perm)
+    out = paddle.index_put(out, (index,), value)
+    out = paddle.transpose(out, perm)
+
+    if inplace:
+        x[:] = out
+        return x
```
Comment (zoooo0820) [inline python/paddle/tensor/manipulation.py:5453]: 此处，如果是inplace，是否可以不要调用clone+setitem赋值，而是直接使用index_put_赋值； 如果非inplace，是否可以不需要额外的clone操作

File: test/legacy_test/test_index_fill.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+
+class TestIndexFillAPI1(TestIndexFillAPIBase):
+    def modify_setting(self):
+        self.dtype_np = 'int64'
+        self.index_type = 'int32'
+        self.x_shape = (10, 15, 10)
+        self.axis = 1
+
+
```
Comment (zoooo0820) [inline test/legacy_test/test_index_fill.py:128]: 补充下complex类型的测试吧

File: test/legacy_test/test_index_fill.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                feed_list = {"x": self.x_np, "index": self.index_np}
+                pd_res = exe.run(
+                    paddle.static.default_main_program(),
+                    feed=feed_list,
+                    fetch_list=[out],
+                )[0]
+                ref_res = compute_index_put_ref(
+                    self.x_np, self.axis, self.index_np, self.value
+                )
+                np.testing.assert_allclose(ref_res, pd_res, atol=1e-5)
```
Comment (zoooo0820) [inline test/legacy_test/test_index_fill.py]: 这几个assert能否使用默认的atol阈值

**Review by zoooo0820 — COMMENTED** [review]
static-check中提示示例代码格式问题，辛苦也再根据报错信息中给出的参考文档修改下

**Comment by Patrick-Star125** [comment]
Done

File: test/legacy_test/test_index_fill.py (hunk 2 above)
Comment (Patrick-Star125) [reply test/legacy_test/test_index_fill.py:128]: index_put不支持complex类型的输入

增加了float16类型的测试

**Comment by zoooo0820** [comment]
> Done

@Patrick-Star125 辛苦再看下static-check中显示的示例代码格式，以及单测中atol阈值设置这两个问题呢

File: test/legacy_test/test_index_fill.py (hunk 3 above)
Comment (Patrick-Star125) [reply test/legacy_test/test_index_fill.py]: Done

**Comment by Patrick-Star125** [comment]
@zoooo0820 似乎static-check中报错不太明确，我主要参考这个[PR](https://github.com/PaddlePaddle/Paddle/pull/57295)的写法，样例代码在我线下都是可以正常执行的

**Comment by Patrick-Star125** [comment]
@zoooo0820 似乎static-check中报错不太明确，我主要参考这个[PR](https://github.com/PaddlePaddle/Paddle/pull/57295)的写法，样例代码在我线下都是可以正常执行的

**Comment by sunzhongkai588** [comment]
> @zoooo0820 似乎static-check中报错不太明确，我主要参考这个[PR](https://github.com/PaddlePaddle/Paddle/pull/57295)的写法，样例代码在我线下都是可以正常执行的

@Patrick-Star125 
- `static-check` 代码示例没通过，报以下错，应该是没 `import numpy`。请尽量不使用 numpy，直接创建paddle tensor。

image

- 另外，示例代码检查默认是cpu环境，输出需要和cpu环境下一致。

详细规范请务必参考 [python文档示例代码规范](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/style_guide_and_references/code_example_writing_specification_cn.html)

**Comment by Patrick-Star125** [comment]
Done

**Review by zoooo0820 — APPROVED** [review]
LGTM

File: python/paddle/tensor/manipulation.py [hunk 4]
```diff
@@ -5166,3 +5166,104 @@ def unfold(x, axis, size, step, name=None):
...
+        value = paddle.to_tensor(value, dtype=x.dtype)
+    else:
+        if len(value.shape) > 0:
+            raise ValueError("value must be scalar or 0-D tensor")
+
+    x_dim = len(x.shape)
+    if axis < 0:
+        axis = axis + x_dim
+
+    if not (isinstance(axis, int)) or (axis > x_dim - 1) or axis < -x_dim:
```
Comment (jeff41404) [inline python/paddle/tensor/manipulation.py:5438]: The negative `axis` has been processed in L5183 above, so the judgment condition here should be `axis < 0` not `axis < -x_dim`.

File: python/paddle/tensor/manipulation.py [hunk 5]
```diff
@@ -5166,3 +5166,104 @@ def unfold(x, axis, size, step, name=None):
...
+@inplace_apis_in_dygraph_only
+def index_fill_(x, index, axis, value, name=None):
+    """
+    Fill the elements of the input tensor with value by the spcific axis and index.
+
+    Args:
+        x (Tensor) : The Destination Tensor. Supported data types are int32, int64, float16, float32, float64.
+        axis (int): The dimension along which to index.
+        index (Tensor): The 1-D Tensor containing the indices to index.
+            The data type of ``index`` must be int32 or int64.
```
Comment (jeff41404) [inline python/paddle/tensor/manipulation.py]: The order of parameters in the document should be consistent with the order of parameters in the function.

File: python/paddle/tensor/manipulation.py (hunk 1 above)
Comment (jeff41404) [reply python/paddle/tensor/manipulation.py:5453]: the  implementation solution in [rfc](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20230914_api_design_for_index_fill.md) shoule be also changed.

File: test/legacy_test/test_index_fill.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+import numpy as np
+
+import paddle
+from paddle.base import Program
+
+paddle.enable_static()
+
+
+def compute_index_put_ref(x, axis, index, value):
```
Comment (jeff41404) [inline test/legacy_test/test_index_fill.py]: `compute_index_fill_ref` is more suitable?

Comment (jeff41404) [reply test/legacy_test/test_index_fill.py]: Is the logic of `compute_index_put_ref` too complex? and it is better to use Numpy's fancy indexing, eg, `x=np.transpose(x, perm); x[index] = value; x=np.transpose(x, perm)` be simpler and easier to understand?

File: python/paddle/tensor/manipulation.py (hunk 1 above)
Comment (Patrick-Star125) [reply python/paddle/tensor/manipulation.py:5453]: current implementation is same with [rfc API design](https://github.com/PaddlePaddle/community/pull/621/files#diff-c670eea33ea5de8561217be758cd152235251086d31135bf65b7abb3b09fdad8R230-R240) already

File: python/paddle/tensor/manipulation.py (hunk 5 above)
Comment (Patrick-Star125) [reply python/paddle/tensor/manipulation.py]: Done

File: python/paddle/tensor/manipulation.py (hunk 4 above)
Comment (Patrick-Star125) [reply python/paddle/tensor/manipulation.py:5438]: Done

File: test/legacy_test/test_index_fill.py (hunk 6 above)
Comment (Patrick-Star125) [reply test/legacy_test/test_index_fill.py]: Done

File: python/paddle/tensor/manipulation.py (hunk 1 above)
Comment (jeff41404) [reply python/paddle/tensor/manipulation.py:5453]: in [rfc](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20230914_api_design_for_index_fill.md?plain=1#L236)  this line: `out = paddle.clone(x)` at else branch need to be delete ?

**Review by jeff41404 — DISMISSED** [review]
LGTM

File: python/paddle/tensor/manipulation.py [hunk 7]
```diff
@@ -5181,3 +5181,104 @@ def unfold(x, axis, size, step, name=None):
...
+        out = paddle.transpose(x, perm)
+        out = paddle.index_put(out, (index,), value)
+        out = paddle.transpose(out, perm)
+        return out
+
+
+def index_fill(x, index, axis, value, name=None):
+    """
+    Outplace version of ``index_fill_`` API, the output Tensor will be inplaced with input ``x``.
+    Please refer to :ref:`api_paddle_index_fill`.
```
Comment (sunzhongkai588) [inline python/paddle/tensor/manipulation.py]:
```suggestion
    Please refer to :ref:`api_paddle_index_fill_`.
```

**Comment by luotao1** [comment]
请解决下冲突以及codestyle流水线的问题

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs

**Review by sunzhongkai588 — COMMENTED** [review]
LGTM～

**Comment by co63oc** [comment]
```
import paddle
x = paddle.eye(num_rows=3, num_columns=4)
indices = paddle.to_tensor(data=[0, 2])
value = -1
result = x.index_fill(axis=1, index=indices, value=value)
print(result)

result = x.index_fill_(axis=1, index=indices, value=value)
print(result)
print(x)
```

```
Tensor(shape=[3, 4], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[-1.,  0., -1.,  0.],
        [-1.,  1., -1.,  0.],
        [-1.,  0., -1.,  0.]])
Tensor(shape=[3, 4], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[-1., -1., -1., -1.],
        [ 0.,  1.,  0.,  0.],
        [-1., -1., -1., -1.]])
Tensor(shape=[3, 4], dtype=float32, place=Place(gpu:0), stop_gradient=True,
       [[-1., -1., -1., -1.],
        [ 0.,  1.,  0.,  0.],
        [-1., -1., -1., -1.]])
```

The index_fill and index_fill_ versions have different calculation results

@Patrick-Star125

**Comment by Patrick-Star125** [comment]
@co63oc [this pr](https://github.com/PaddlePaddle/Paddle/pull/59863) has been submitted for this bug, thank you for your feedback

**Pull request merged**